crates/integration_test/src/tests/lifecycle/checkpoint.rs

```diff
@@ -0,0 +1,168 @@
+use super::get_result_from_output;
+use crate::utils::get_runtime_path;
+use crate::utils::test_utils::State;
+use crate::utils::{create_temp_dir, generate_uuid};
+use anyhow::anyhow;
+use std::path::Path;
+use std::process::{Command, Stdio};
+use test_framework::TestResult;
+
+// Simple function to figure out the PID of the first container process
+fn get_container_pid(project_path: &Path, id: &str) -> Result<i32, TestResult> {
+    let res_state = match Command::new(get_runtime_path())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .arg("--root")
+        .arg(project_path.join("runtime"))
+        .arg("state")
+        .arg(id)
+        .spawn()
+        .expect("failed to execute state command")
+        .wait_with_output()
+    {
+        Ok(o) => o,
+        Err(e) => {
+            return Err(TestResult::Failed(anyhow!(
+                "error getting container state {}",
+                e
+            )))
+        }
+    };
+    let stdout = match String::from_utf8(res_state.stdout) {
+        Ok(s) => s,
+        Err(e) => {
+            return Err(TestResult::Failed(anyhow!(
+                "failed to parse container stdout {}",
+                e
+            )))
+        }
+    };
+    let state: State = match serde_json::from_str(&stdout) {
+        Ok(v) => v,
+        Err(e) => {
+            return Err(TestResult::Failed(anyhow!(
+                "error in parsing state of container: stdout : {}, parse error : {}",
+                stdout,
+                e
+            )))
+        }
+    };
+
+    Ok(match state.pid {
+        Some(p) => p,
+        _ => -1,
+    })
+}
+
+// CRIU requires a minimal network setup in the network namespace
+fn setup_network_namespace(project_path: &Path, id: &str) -> Result<(), TestResult> {
+    let pid = get_container_pid(project_path, id)?;
+
+    if let Err(e) = Command::new("nsenter")
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .arg("-t")
+        .arg(format!("{}", pid))
+        .arg("-a")
+        .args(vec!["/bin/ip", "link", "set", "up", "dev", "lo"])
+        .spawn()
+        .expect("failed to exec ip")
+        .wait_with_output()
+    {
+        return Err(TestResult::Failed(anyhow!(
+            "error setting up network namespace {}",
+            e
+        )));
+    }
+
+    Ok(())
+}
+
+fn checkpoint(
+    project_path: &Path,
+    id: &str,
+    args: Vec<&str>,
+    work_path: Option<&str>,
+) -> TestResult {
+    if let Err(e) = setup_network_namespace(project_path, id) {
+        return e;
+    }
+
+    let temp_dir = match create_temp_dir(&generate_uuid()) {
+        Ok(td) => td,
+        Err(e) => {
+            return TestResult::Failed(anyhow::anyhow!(
+                "Failed creating temporary directory {:?}",
```

Comment:
Can I ask you to make all error messages start lowercase letters?
```suggestion
                "failed creating temporary directory {:?}",
```